bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/CommonWebSocketServlet.java

```diff
@@ -0,0 +1,149 @@
+        webSocketServletFactory.getPolicy().setIdleTimeout(10000);
+        webSocketServletFactory.setCreator(new CommonWebSocketCreator());
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addWebSocketHandler(WebSocketHandler wsConnectionHandler) {
+        this.connectionHandlers.put(wsConnectionHandler.getId(), wsConnectionHandler);
+    }
+
+    protected void removeWebSocketHandler(WebSocketHandler wsConnectionHandler) {
+        this.connectionHandlers.remove(wsConnectionHandler.getId());
+    }
+
+    private class CommonWebSocketCreator implements WebSocketCreator {
+        private final Logger logger = LoggerFactory.getLogger(CommonWebSocketCreator.class);
+
+        @Override
+        public @Nullable Object createWebSocket(@Nullable ServletUpgradeRequest servletUpgradeRequest,
+                @Nullable ServletUpgradeResponse servletUpgradeResponse) {
+            if (servletUpgradeRequest == null) {
+                return null;
+            }
+            if (isAuthorizedRequest(servletUpgradeRequest)) {
+                String requestPath = servletUpgradeRequest.getRequestURI().getPath();
+                String handlerPrefix = SERVLET_PATH + "/";
+                boolean useDefaultHandler = requestPath.equals(handlerPrefix) || !requestPath.startsWith(handlerPrefix);
+                WebSocketHandler wsHandler;
+                if (!useDefaultHandler) {
+                    String handlerId = requestPath.substring(handlerPrefix.length());
+                    wsHandler = connectionHandlers.get(handlerId);
+                    if (wsHandler == null) {
+                        logger.warn("Missing WebSocket handler for path {}", handlerId);
+                        return null;
+                    }
+                } else {
+                    wsHandler = connectionHandlers.get(DEFAULT_HANDLER_ID);
+                    if (wsHandler == null) {
+                        logger.warn("Default WebSocket handler is missing");
+                        return null;
+                    }
+                }
+                logger.debug("New connection handled by {}", wsHandler.getId());
+                return wsHandler.createWebSocket(servletUpgradeRequest, servletUpgradeResponse);
+            } else {
+                logger.warn("Unauthenticated request to create a websocket from {}.",
+                        servletUpgradeRequest.getRemoteAddress());
+            }
+            return null;
+        }
+
+        private boolean isAuthorizedRequest(ServletUpgradeRequest servletUpgradeRequest) {
+            Map<String, List<String>> parameterMap = servletUpgradeRequest.getParameterMap();
+            List<String> accessToken = parameterMap.getOrDefault("accessToken", List.of());
+            SecurityContext securityContext = null;
+            if (accessToken.isEmpty() && authFilter.isImplicitUserRole(servletUpgradeRequest.getHttpServletRequest())) {
```

Comment:
This code builds assumptions on a top of another component configuration. It would be better to centralize that part elsewhere.

Reply:
Yes, that answers the question on the previous comment.

I also saw this comment:

> This dependency brings you a lot of stuff, making websocket part embedding a lot of jaxrs dependencies. From this point of view, if you collapse dependencies you could try to do websockets at jaxrs level. it will be slower, but will keep things consistent.

Do you think that creating a new bundle with just the required stuff could be a good idea?

Reply:
You construct a security context in the same way as `AuthFilter` would do thus its better to hide this detail there.

Reply:
Sure, It will look better, I can refactor the code to handle it all in the AuthFilter class.

The initial reason to have it split is that the AccessToken for the websocket can be passed on the query.

For context, the reason for this is that the WebSocket interface exposed in browsers does not allow you to set other headers than "Sec-WebSocket-Protocol". So sending the token in the authentication header is not possible from a browser.

Reply:
Right, in such case you could try to get it around with cookie, but not sure if it will not complicate things even more. :/

Reply:
Yes, normally try to avoid using cookies.

I have added some changes to the AuthFilter so the logic is hidden there in the 'getSecuryContext' method.